packages/rpc/src/adaptors/iframe-child.ts

```diff
@@ -7,12 +7,31 @@ export function fromInsideIframe({targetOrigin = '*'} = {}): MessageEndpoint {
     );
   }
 
+  /**
+   * We wait until the document is ready before advertising to the parent that
+   * communication can commence.
+   *
+   * However, it's possible that the parent isn't listening to messages at this time.
+   * Which can lead to communication never starting.
+   *
+   * Therefore we also wait for the parent to send a message once it's ready to (re)send the
+   * ready message from within the child iframe.
+   */
+
   const {parent} = self;
 
   const ready = () => parent.postMessage('remote-ui::ready', targetOrigin);
 
-  // Listening to `readyState` in iframe, though the child iframe could probably
-  // send a `postMessage` that it is ready to receive messages sooner than that.
+  window.addEventListener('message', (event) => {
+    if (event.source !== parent || document.readyState !== 'complete') {
```

Comment:
Do you know if it's possible for this to happen for the document is ready?

Reply:
Do you mean _before_ the document is ready?

Reply:
I actually had the same question, but assumed that the previous authors have checked this. 

But I'd try to check

Reply:
@lemonmade it seems that this can actually happen. It was hard to reproduce since all our test apps use `type=module` which defers scripts and hence only executes them after the document is loaded.

However, if that's not the case, `document.readyState` can be `loading` under some circumstances when the iframe loads a blocking JS after the file that contains the rpc code.

![image](https://user-images.githubusercontent.com/474248/221988870-41056733-fd9f-4776-8635-c3beb08beaef.png)